movies-manager/src/containers/movie-form/movie-form.tsx

```diff
@@ -0,0 +1,178 @@
+import React, { Dispatch, MouseEvent } from 'react';
+import { connect } from 'react-redux';
+import { faCalendar } from '@fortawesome/free-solid-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { Formik, Form } from 'formik';
+import * as Yup from 'yup';
+import './movie-form.scss';
+import { ICheckboxGenre, IMovie, IStoreState, IMovieInfo } from '../../types/types';
+import { InputControl } from '../../components/input-control/input-control';
+import { SelectControl } from '../../components/select-control/select-control';
+import { createMovieFromServer, editMovieFromServer } from '../../redux/thunks/movies-thunks';
+import { defaultMovie } from './mockDefaultMovie';
+import { allGenres } from '../../utils/variables';
+
+interface ISaveChanges {
+    movie?: IMovie;
+    onSaveChanges: () => void;
+    createMovie(movie: IMovie): void;
+    editMovie(movie: IMovie): void;
+}
+
+type THandleSubmit = (movieInfo: IMovie) => void;
+type TResetForm = () => void;
+type TResetHandlerWrapper = (resetForm: TResetForm) => (e: MouseEvent) => void;
+
+const blockName = 'form-test';
+
+enum FORM_HEADLINES {
+    EDIT_MOVIE = 'Edit Movie',
+    ADD_MOVIE = 'Add Movie',
+}
+
+function MovieForm({
+    movie = defaultMovie,
+    onSaveChanges,
+    createMovie,
+    editMovie
+}: ISaveChanges): JSX.Element {
+    const handleSubmit: THandleSubmit = (movieInfo: IMovie) => {
+        const method = movieInfo.id ? editMovie : createMovie;
+        method(movieInfo);
+        onSaveChanges();
+    };
+
+    const resetMovieForm: TResetHandlerWrapper = (resetForm: TResetForm) => (e: MouseEvent) => {
+        e.preventDefault();
+        e.stopPropagation();
+        resetForm();
+    };
+
+    const availableGenres = Array.from(new Set([...allGenres, ...movie.genres]));
+    const genres = availableGenres.map((genre: string) => ({
+        label: genre,
+        isChecked: movie.genres.includes(genre),
+    }));
+
+    const movieIdField: JSX.Element | undefined = movie.id && <div className={`${blockName}__field-wrapper`}>
+        <div className={`${blockName}__title`}>movie id</div>
+        <div className={`${blockName}__text`}>{movie.id}</div>
+    </div>;
+
+    const icoCalendar: JSX.Element = <FontAwesomeIcon
+        className='input-control__icon--bright'
+        icon={faCalendar}/>;
+
+    return <Formik
+        initialValues={{ ...movie, genres }}
+        validationSchema={Yup.object({
+            title: Yup.string()
+                .min(3, 'Must be 3 characters or more')
+                .required('Required'),
+            release_date: Yup.string()
+                .matches(
+                    /(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])/,
```

Comment:
Find better regex:
![validation_problem_lenkov](https://user-images.githubusercontent.com/20605638/96239279-2b481a00-0fa8-11eb-881b-fb21afab7626.gif)

Reply:
Also, I know this was not part of requirements, but add smth like https://material-ui.com/ru/components/snackbars/ (you can use any library or write your custom component to display error messages).

Reply:
Done